### DISCUSSION TITLE
[SUGGESTION] Add MP3 playback for transcription

### DISCUSSION THREAD

**Discussion by lthoerner** [musescore/MuseScore#11801] [Ideas]
Hey, I had another idea. I've been having a lot of trouble getting the correct timings when transcribing pieces with variable tempos. Because Musescore playback is often not instant after clicking/pressing the play button - except, for some reason, at the start of a score - I have had to resort to playing an audio file in another window, and trying to click the play button at the perfect moment. This wastes a lot of time, as I have to start from the very beginning of the track to make sure everything is lined up correctly (due to the aforementioned playback issue).

So, is it possible to add a feature where I can add in a track that simply can be played along with the other parts (much like just another instrument)? This would save hours of my time and I feel like it's one of those things that people would love but wouldn't realize that they needed initially. I'd be happy to hear what you guys have to say. Thanks!

**Comment by GabeS573** [comment]
A feature like this could also be extended to include playback of audio samples in a score — for example, in marching band field shows, which often include audio samples.

**Comment by jsm222** [comment]
If you are on Apple you can have a look here, https://github.com/jsm222/MuseScore/releases/tag/v0.0.1-em for a simple implementation... The code is here https://github.com/musescore/MuseScore/compare/master...jsm222:external-media..

**Reply by lthoerner** [reply]
I'm not on Apple, but did you literally just implement the feature I asked for? If so you're an absolute chad

**Reply by jsm222** [reply]
Thanks, yes more or less some things are missing though. If you do a loop, boundaries are not adhered to by the external media, and the volume should probably be adjustable.. And probably some other missing features. But it is a proof of concept/starting point.. Just note that on Linux the first playback of the external media starts late, but stop rewind and play, the sound starts timely.

**Reply by cbjeukendrup** [reply]
Nice experiment! I don't expect though that your branch will be merged 'as is'. Mainly because we don't want to add another Qt module as a dependency. But also because this feature should be implemented in a much higher layer of the playback engine. See `mu::audio::SoundTrack` in src/framework/audio/internal/worker/track.h. It looks like a lot of groundwork for this feature is already in place.

**Reply by jsm222** [reply]
It sure does, it seems  AudioStream::loadFille could do some tricks, and that would probably put the external media track into the mixer as well. I will have a look at it..

**Comment by jsm222** [comment]
https://github.com/musescore/MuseScore/compare/master...jsm222:external-wav is a stub that starts implementing for wav only. Place a file in your home called mscoretest.wav and it playes a long with the other instruments. loop and seek supported. Playback starts right away, and I have not found a way to stop or mute it.. I do not really understand the event system yet, or the in and output params fully, but  it is a starting point. To convert from mp3 do for instance 
```ffmpeg -i test.mp3 -f wav -acodec pcm_f32le -ar 48000 -ac 2 ~/mscoretest.wav```

**Comment by lunabunn** [comment]
bump. Ideally this would come with two features:
- Volume adjustment, as mentioned above
- Playback offset in case the audio file isn't "on beat"

**Comment by juniorerico** [comment]
any progress on this topic? I'm using @jsm222 version and it is great as a start, exactly what I wanted!

**Reply by cbjeukendrup** [reply]
It might be a good idea to create a fresh feature request issue for this: https://github.com/musescore/MuseScore/issues/new?assignees=&labels=Feature+Request&projects=&template=feature_request.yml

I see that the original feature request got closed a few years ago and the discussion was moved here, but things have changed a bit since then. Now that the first thoughts about MuseScore Studio 5.0 are appearing, it might be good to have this request listed somewhere. 

That's not to imply that this will be implemented soon though. The first step is that it will need a proper design from the design team.
